File: src/main/scala/models/Job.scala
package me.amanj.greenish.models

import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.time.ZoneId
import io.circe.{Encoder, Decoder, HCursor, Json}
import io.circe.generic.extras.semiauto.{deriveConfiguredDecoder, deriveConfiguredEncoder}

case class Job(
  jobId: Int,
  name: String,
  prometheusId: String,
  cmd: String,
  timePattern: String,
  frequency: CheckFrequency,
  periodCheckOffset: Int,
  timezone: ZoneId,
  lookback: Int,
  startAt: Long,
  alertLevels: AlertLevels,
  env: Seq[(String, String)]
) {
  val timeFormat = DateTimeFormatter.ofPattern(timePattern)
}

object Job {
  implicit val zoneIdEncoder: Encoder[ZoneId] =
    new Encoder[ZoneId] {
      final def apply(zid: ZoneId): Json = Json.obj(
        ("zone_id", Json.fromString(zid.getId))
      )
    }
  implicit val zoneIdDecoer: Decoder[ZoneId] = new Decoder[ZoneId] {
    final def apply(c: HCursor): Decoder.Result[ZoneId] =
      for {
        zoneId <- c.downField("zone_id").as[String]
      } yield ZoneId.of(zoneId)
  }

  implicit val checkEntryDecoder: Decoder[Job] = deriveConfiguredDecoder
  implicit val checkEntryEncoder: Encoder[Job] = deriveConfiguredEncoder
}

File: src/test/resources/json-samples/job.js
module.exports = {
  "testObject": {
    {
      "job": {
        "job_id": 0,
        "name": "Raw event logs",
        "prometheus_id": "reporting_hourly_jobs_raw_event_logs",
        "cmd": "/tmp/check /tmp",
        "time_pattern": "yyyy-MM-dd-HH",
        "frequency": "hourly",
        "period_check_offset": 1,
        "timezone": {
          "zone_id": "UTC"
        },
        "lookback": 6,
        "alert_levels": {
          "great": 0,
          "normal": 1,
          "warn": 2,
          "critical": 3
        },
        "env": []
      },
      "updated_at": 1595126730945,
      "period_health": [
        {
          "period": "2020-07-18-20",
          "ok": true
        },
        {
          "period": "2020-07-18-21",
          "ok": false
        },
        {
          "period": "2020-07-18-22",
          "ok": true
        },
        {
          "period": "2020-07-18-23",
          "ok": true
        },
        {
          "period": "2020-07-19-00",
          "ok": false
        },
        {
          "period": "2020-07-19-01",
          "ok": false
        }
      ]
    }
  }
}
